README.md
[![CodeFactor](https://www.codefactor.io/repository/github/danitetus/pushover-js/badge)](https://www.codefactor.io/repository/github/danitetus/pushover-js) ![Build test](https://github.com/danitetus/pushover-js/workflows/Build%20test/badge.svg)

# pushover-js v1.3.0

A Node JS module written in Typescript to use with Pushover notifications app

---

## Breaking changes from v0.x.x

This entire library has been rewritten to be based on promises.

The code has been restructured, and the methods changed to make it more complete and versatile.

If you were using a version of the 0.x.x branch, please check the documentation to adapt your code.

---

## Installation

```bash
npm install --save pushover-js
```

---

## Documentation

Please read the complete documentation in the [docs](docs/globals.md) folder

## Usage

### Promises

```javascript
/*
        JAVASCRIPT
*/

// Load the module
var Pushover = require( 'pushover-js').Pushover;

// Instance
const pushover = new Pushover('user', 'token')

// Simple notification (without personalization)
pushover
    .send('My title', 'My message')
        .then(console.log)
        .catch(console.error)

// Custom notification
pushover
    // Setting the sound
    .setSound('cashregister')
    // If the priority is 2, you have to provide an expire time and a retry time (see the official API for more info)
    .setPriority(2, 60, 30)
    // Set an url to access on tap
    .setUrl('https://www.github.com/danitetus/pushover-js', 'Pushover-JS github')
    // Set custom timestamp (must to be unix timestamp, not javascript time!!!)
    .setTimestamp(Math.floor(new Date(2020, 2, 1).getTime() / 1000))
    // Add a file
    .setAttachment('image.jpg', 'path/to/your/file.jpg')
    // Mark message as html
    .setHtml()
    // Send
    .send('Submit an issue', 'Click on the link to submit an issue')
  .then((msj) => {
    console.log(msj)
  })
  .catch((e) => {
    console.error(e)
  })
```

```typescript
/*
        TYPESCRIPT
*/

// Load the module
import { Pushover } from 'pushover-js'

// Instance
const pushover = new Pushover('user', 'token')

// Simple notification (without personalization)
pushover
    .send('My title', 'My message')
        .then(console.log)
        .catch(console.error)

// Custom notification
pushover
    // Setting the sound
    .setSound('cashregister')
    // If the priority is 2, you have to provide an expire time and a retry time (see the official API for more info)
    .setPriority(2, 60, 30)
    // Set an url to access on tap
    .setUrl('https://www.github.com/danitetus/pushover-js', 'Pushover-JS github')
    // Set custom timestamp (must to be unix timestamp, not javascript time!!!)
    .setTimestamp(Math.floor(new Date(2020, 2, 1).getTime() / 1000))
    // Add a file
    .setAttachment('image.jpg', 'path/to/your/file.jpg')
    // Mark message as html
    .setHtml()
    // Send
    .send('Submit an issue', 'Click on the link to submit an issue')
  .then((msj) => {
    console.log(msj)
  })
  .catch((e) => {
    console.error(e)
  })
```

### Async/await

```typescript
// Load the module
import { Pushover } from 'pushover-js'


// Use this inside an async function
const sendNotification = async () => {
    // Instance
    const pushover = new Pushover('user', 'token')

    // Custom notification
    pushover
        // Setting the sound
        .setSound('cashregister')
        // If the priority is 2, you have to provide an expire time and a retry time (see the official API for more info)
        .setPriority(2, 60, 30)
        // Add a file
        .setAttachment('image.jpg', 'path/to/your/file.jpg')
        // Set an url to access on tap
        .setUrl('https://www.github.com/danitetus/pushover-js', 'Pushover-JS github')

    try {
        const response = await Pushover.send('Submit an issue', 'Click on the link to submit an issue')
        console.log(response)
    } catch (error) {
        console.error(error)
    }
}

```

For any suggestion or issue, please go to the [issues section](https://github.com/danitetus/pushover-js/issues) on the [Github repository](https://github.com/danitetus/pushover-js)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.0] - 2020-12-08
### Added
- Now you can attach a file in your notifications

## [1.2.0] - 2020-12-08
### Added
- Allow marking messages as html compliant

## [1.1.1] - 2020-12-08
### Added
- Timestamp setter for personalize the time when notification was created

### Fixed
- Successfully requests doesn't resolve the promise and never returns the response object


lib/request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const https = require('https');
class ResponseError {
    constructor(message, statusCode) {
        this.name = 'Response error';
        this.message = message;
        this.statusCode = statusCode;
    }
}
const processParam = async (name, value) => {
    return `\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`;
};
const processFile = async (name, filePath, boundary) => {
    let type = 'application/octet-stream';
    if (/\.(jpg|jpeg)$/i.test(name)) {
        type = 'image/jpeg';
    }
    else if (/\.png$/i.test(name)) {
        type = 'image/png';
    }
    else if (/\.gif$/i.test(name)) {
        type = 'image/gif';
    }
    else if (/\.mp3$/i.test(name)) {
        type = 'audio/mpeg';
    }
    else if (/\.mp4$/i.test(name)) {
        type = 'video/mp4';
    }
    const data = await fs.promises.readFile(filePath);
    return Buffer.concat([
        Buffer.from(`\r\n--${boundary}\r\nContent-Disposition: form-data; name="attachment"; filename="${name}"\r\nContent-type: ${type}\r\n\r\n`),
        data,
        Buffer.from(`\r\n`)
    ]);
};
const responseCallback = (res) => {
    return new Promise((resolve, reject) => {
        let responseData = '';
        res.on('data', (data) => {
            responseData += data.toString();
        });
        res.on('end', () => {
            if (res.statusCode && res.statusCode !== 200) {
                const error = new ResponseError(responseData, res.statusCode);
                reject(error);
            }
            else {
                const returnData = {
                    headers: res.headers,
                    data: responseData,
                    statusCode: typeof res.statusCode === 'number' ? res.statusCode : 500
                };
                resolve(returnData);
            }
        });
    });
};
exports.default = {
    get: async (options) => {
        return new Promise((resolve, reject) => {
            options = options || {};
            options = { ...options, ...{ method: 'GET' } };
            const req = https.request(options, (res) => {
                responseCallback(res).then(resolve).catch(reject);
            });
            req.end();
        });
    },
    post: async (options, postData) => {
        return new Promise(async (resolve, reject) => {
            const boundary = `------------------------${Math.random().toString(36).substring(2)}`;
            options.headers = {};
            options.headers['Content-type'] = `multipart/form-data; boundary=${boundary}`;
            options.method = 'POST';
            let data = [];
            let fileData = Buffer.from('', 'utf-8');
            for (const param in postData) {
                if (param === 'file') {
                    fileData = await processFile(postData[param].name, postData[param].filePath, boundary);
                }
                else {
                    const payload = await processParam(param, postData[param]);
                    data.push(payload);
                }
            }
            let body;
            try {
                body = Buffer.concat([
                    Buffer.from(`--${boundary}`),
                    Buffer.from(data.join(`--${boundary}`), 'utf-8'),
                    fileData,
                    Buffer.from(`--${boundary}--\r\n`, 'utf-8')
                ]);
            }
            catch (e) {
                reject(e);
                return;
            }
            options.headers['Content-length'] = body.length;
            const req = https.request(options, (res) => {
                responseCallback(res).then(resolve).catch(reject);
            });
            req.write(body);
            req.end();
        });
    }
};
//# sourceMappingURL=request.js.map

lib/pushover.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pushover = void 0;
const request_1 = __importDefault(require("./request"));
class Pushover {
    constructor(user, token) {
        this._hostname = 'api.pushover.net';
        this._path = '/1/messages.json';
        this._notification = {
            user: '',
            token: '',
            title: '',
            message: '',
            sound: 'pushover',
            priority: 0,
            expire: 0,
            retry: 0
        };
        this._notification.user = user;
        this._notification.token = token;
    }
    setDevice(device) {
        this._notification.device = device;
        return this;
    }
    setHtml() {
        this._notification.html = 1;
        return this;
    }
    setTitle(title) {
        this._notification.title = title;
        return this;
    }
    setMessage(message) {
        this._notification.message = message;
        return this;
    }
    setSound(sound) {
        this._notification.sound = sound;
        return this;
    }
    setAttachment(name, filePath) {
        this._notification.file = {
            name,
            filePath
        };
        return this;
    }
    setPriority(priority, expire, retry) {
        if (priority < -2 || priority > 2) {
            console.error('Pushover notification priority have to be from -2 to 2');
            return this;
        }
        this._notification.priority = priority;
        if (priority === 2) {
            this._notification.expire = typeof expire === 'number' ? expire : 10800;
            this._notification.retry = typeof retry === 'number' ? retry : 3600;
        }
        return this;
    }
    setUrl(url, title) {
        this._notification.url = url;
        if (title) {
            this._notification.url_title = title;
        }
        return this;
    }
    setTimestamp(timestamp) {
        this._notification.timestamp = timestamp;
        return this;
    }
    async send(title, message) {
        if (title) {
            this._notification.title = title;
        }
        if (message) {
            this._notification.message = message;
        }
        const options = {
            port: 443,
            host: this._hostname,
            path: this._path,
        };
        return request_1.default.post(options, this._notification);
    }
}
exports.Pushover = Pushover;
//# sourceMappingURL=pushover.js.map

src/pushover.ts
import Request from './request'
import { IResponse } from './request'


type Priority = -2 | -1 | 0 | 1 | 2

export interface INotificationData {
  user: string;
  token: string;
  device?: string;
  url?: string;
  url_title?: string;
  timestamp?: number;
  html?: 1;
  title: string;
  message: string;
  sound: Sound;
  priority: number;
  expire: number;
  retry: number;
  file?: {
    name: string;
    filePath: string;
  }
}

type Sound = 'pushover' |
  'bike' |
  'bugle' |
  'cashregister' |
  'classical' |
  'cosmic' |
  'falling' |
  'gamelan' |
  'incoming' |
  'intermission' |
  'magic' |
  'mechanical' |
  'pianobar' |
  'siren' |
  'spacealarm' |
  'tugboat' |
  'alien' |
  'climb' |
  'persistent' |
  'echo' |
  'updown' |
  'vibrate' |
  'none'

export class Pushover {
  private _hostname: string = 'api.pushover.net'
  private _path: string = '/1/messages.json'

  private _notification: INotificationData = {
    user: '',
    token: '',
    title: '',
    message: '',
    sound: 'pushover',
    priority: 0,
    expire: 0,
    retry: 0
  }

  constructor(user: string, token: string) {
    this._notification.user = user
    this._notification.token = token
  }

  public setDevice(device: string): Pushover {
    this._notification.device = device
    return this
  }

  public setHtml(): Pushover {
    this._notification.html = 1
    return this
  }

  public setTitle(title: string): Pushover {
    this._notification.title = title
    return this
  }

  public setMessage(message: string): Pushover {
    this._notification.message = message
    return this
  }

  public setSound(sound: Sound): Pushover {
    this._notification.sound = sound
    return this
  }

  public setAttachment(name: string, filePath: string): Pushover {
    this._notification.file = {
      name,
      filePath
    }
    return this
  }

  public setPriority(priority: Priority, expire?: number, retry?: number): Pushover {
    if (priority < -2 || priority > 2) {
      console.error('Pushover notification priority have to be from -2 to 2')
      return this
    }
    this._notification.priority = priority

    if (priority === 2) {
      this._notification.expire = typeof expire === 'number' ? expire : 10800
      this._notification.retry = typeof retry === 'number' ? retry : 3600
    }
    return this
  }

  public setUrl(url: string, title?: string): Pushover {
    this._notification.url = url
    if (title) {
      this._notification.url_title = title
    }
    return this
  }

  public setTimestamp(timestamp: number): Pushover {
    this._notification.timestamp = timestamp
    return this
  }

  public async send(title?: string, message?: string): Promise<IResponse> {
    if (title) {
      this._notification.title = title
    }

    if (message) {
      this._notification.message = message
    }

    const options = {
      port: 443,
      host: this._hostname,
      path: this._path,
    }

    return Request.post(options, this._notification)
  }
}


lib/pushover.d.ts
import { IResponse } from './request';
declare type Priority = -2 | -1 | 0 | 1 | 2;
export interface INotificationData {
    user: string;
    token: string;
    device?: string;
    url?: string;
    url_title?: string;
    timestamp?: number;
    html?: 1;
    title: string;
    message: string;
    sound: Sound;
    priority: number;
    expire: number;
    retry: number;
    file?: {
        name: string;
        filePath: string;
    };
}
declare type Sound = 'pushover' | 'bike' | 'bugle' | 'cashregister' | 'classical' | 'cosmic' | 'falling' | 'gamelan' | 'incoming' | 'intermission' | 'magic' | 'mechanical' | 'pianobar' | 'siren' | 'spacealarm' | 'tugboat' | 'alien' | 'climb' | 'persistent' | 'echo' | 'updown' | 'vibrate' | 'none';
export declare class Pushover {
    private _hostname;
    private _path;
    private _notification;
    constructor(user: string, token: string);
    setDevice(device: string): Pushover;
    setHtml(): Pushover;
    setTitle(title: string): Pushover;
    setMessage(message: string): Pushover;
    setSound(sound: Sound): Pushover;
    setAttachment(name: string, filePath: string): Pushover;
    setPriority(priority: Priority, expire?: number, retry?: number): Pushover;
    setUrl(url: string, title?: string): Pushover;
    setTimestamp(timestamp: number): Pushover;
    send(title?: string, message?: string): Promise<IResponse>;
}
export {};
//# sourceMappingURL=pushover.d.ts.map

src/request.ts
import { IncomingMessage } from 'http'
import { INotificationData } from './pushover'
import * as fs from 'fs'
const https = require('https')

export interface IResponse {
  headers: any
  data: string
  statusCode: number
}

class ResponseError implements Error {
  public name: string
  public message: string
  public statusCode: number

  constructor(message: string, statusCode: number) {
    this.name = 'Response error'
    this.message = message
    this.statusCode = statusCode
  }
}

const processParam = async (name: string, value: string): Promise<string> => {
  return `\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`
}

const processFile = async (name: string, filePath: string, boundary: string): Promise<Buffer> => {
  let type = 'application/octet-stream'
  if (/\.(jpg|jpeg)$/i.test(name)) {
    type = 'image/jpeg'
  } else if (/\.png$/i.test(name)) {
    type = 'image/png'
  } else if (/\.gif$/i.test(name)) {
    type = 'image/gif'
  } else if (/\.mp3$/i.test(name)) {
    type = 'audio/mpeg'
  } else if (/\.mp4$/i.test(name)) {
    type = 'video/mp4'
  }
  const data = await fs.promises.readFile(filePath)
  return Buffer.concat([
      Buffer.from(`\r\n--${boundary}\r\nContent-Disposition: form-data; name="attachment"; filename="${name}"\r\nContent-type: ${type}\r\n\r\n`),
      data,
      Buffer.from(`\r\n`)
  ])
}

const responseCallback = (res: IncomingMessage): Promise<IResponse> => {
  return new Promise((resolve, reject) => {
    let responseData: string = ''
    res.on('data', (data) => {
      responseData += data.toString()
    })

    res.on('end', () => {
      if (res.statusCode && res.statusCode !== 200) {
        const error = new ResponseError(responseData, res.statusCode)
        reject(error)
      } else {
        const returnData: IResponse = {
          headers: res.headers,
          data: responseData,
          statusCode: typeof res.statusCode === 'number' ? res.statusCode : 500
        }
        resolve(returnData)
      }
    })
  })
}


export default {
  get: async (options?: any): Promise<IResponse> => {
    return new Promise((resolve, reject) => {
      options = options || {}
      options = { ...options, ...{ method: 'GET' } }
      const req = https.request(options, (res: IncomingMessage) => {
        responseCallback(res).then(resolve).catch(reject)
      })
      req.end()
    })
  },
  post: async (options: any, postData: INotificationData | any): Promise<IResponse> => {
    return new Promise(async (resolve, reject) => {
      const boundary = `------------------------${Math.random().toString(36).substring(2)}`
      options.headers = {}
      options.headers['Content-type'] = `multipart/form-data; boundary=${boundary}`
      options.method = 'POST'
      let data: string[] = []
      let fileData: Buffer = Buffer.from('', 'utf-8')

      for (const param in postData) {
        if (param === 'file') {
          fileData = await processFile(postData[param].name, postData[param].filePath, boundary)
        } else {
          const payload = await processParam(param, postData[param])
          data.push(payload)
        }
      }

      let body: Buffer
      try {
        body = Buffer.concat([
            Buffer.from(`--${boundary}`),
            Buffer.from(data.join(`--${boundary}`), 'utf-8'),
            fileData,
            Buffer.from(`--${boundary}--\r\n`, 'utf-8')
        ])
      } catch (e) {
        reject(e)
        return
      }

      options.headers['Content-length'] = body.length

      const req = https.request(options, (res: IncomingMessage) => {
        responseCallback(res).then(resolve).catch(reject)
      })

      req.write(body)
      req.end()
    })
  }
}
